WIP: api using ffig

petr-tik opened the pull request:
I've resumed work (it's been a long delay - I know) on the API for the ffig-explorer website. 

The main FFIG project is built with source files in mind. Hence most methods read files from disk into memory and then build translation units etc. The way I imagine the server is parsing source files from requests and acting on them in memory. At the moment this code throws an exception in build_model_from_source, which passes the path_to_file to the clang function

``` 
tu = clang.cindex.TranslationUnit.from_source(path_to_source, ...
```

I can either roll my own clang.cindex.TranslationUnit or can add code into main ffig repo, which can expose a small API to take code as a long string in-memory string and generate bindings.

petr-tik committed:
Using new test input

petr-tik committed:
Update python and make sure pythonenv allows import

petr-tik committed:
Re-writing the API

Idea: work out the number of lines of code that the ffig-explorer
web-app needs to turn an input file string into a binding that can be
posted back to the client

petr-tik commented:
This isn't about merging - I want to get your thoughts on the direction of work first.

petr-tik committed:
Add a make command to run the container with API server

ajbennieston commented on Dockerfile:
```diff
@@ -16,17 +16,17 @@ RUN pip2 install --upgrade pip && \
 # separate layer to use docker cache
 RUN cd /home && \
     # clone and init the git submodules
```
Can probably remove this comment; the command below is fairly self-explanatory.

ajbennieston commented on Makefile:
```diff
@@ -1,6 +1,10 @@
 build:
```
Consider adding dependencies on the Dockerfile and any contents that are brought in via `ADD` or `COPY`. This will help Make to determine when it needs to re-run the command.

One approach to ensure this would be to have an arrangement like:

```make
.docker-build: Dockerfile ...
    docker build -t ffig/web-base:latest .
    touch .docker-build

build: .docker-build

run: build
    docker run -p 5000:5000 -it ffig/web-base:latest
```

ajbennieston commented on requests_at_explorer.py:
```diff
@@ -1,31 +1,31 @@
 r = requests.post(
     "http://127.0.0.1:5000/api/gen_bindings_from_tu", data=payload)
-print(r.text)
+if r.status_code == 200:
```
```python
if r.status_code == requests.codes.ok:
```
Avoid magic numbers in code :)
http://docs.python-requests.org/en/master/api/#status-code-lookup

ajbennieston commented:
The `from_source` method is capable of handling a TU from an in-memory string. We just need to make FFIG accept both and pass either the `filename` or `unsaved_files` as appropriate.

`clang/cindex.py:2730`:
```python
    @classmethod
    def from_source(cls, filename, args=None, unsaved_files=None, options=0,
                    index=None):
        """Create a TranslationUnit by parsing source.

        This is capable of processing source code both from files on the
        filesystem as well as in-memory contents.

        Command-line arguments that would be passed to clang are specified as
        a list via args. These can be used to specify include paths, warnings,
        etc. e.g. ["-Wall", "-I/path/to/include"].

        In-memory file content can be provided via unsaved_files. This is an
        iterable of 2-tuples. The first element is the str filename. The
        second element defines the content. Content can be provided as str
        source code or as file objects (anything with a read() method). If
        a file object is being used, content will be read until EOF and the
        read cursor will not be reset to its original position.
        ...
        """
```

petr-tik commented:
Great - sorry I didn't add my question last night. 

Should I make a new method in ffig/ffig.py that takes an in-memory string and choice of binding and returns 1 in-memory string of auto-generated code?

petr-tik commented:
Will address the code review suggestions tomorrow as well

ajbennieston commented:
@petr-tik Unfortunately I think it's a little more complicated than that...

Some bindings generate more than one output file, and the current generator mechanism is still somewhat tied to file output.

Having just refreshed my memory on the FFIG.py source, I think we will need to extend `build_model_from_source` (or write another variant of it) to handle in-memory files, then modify `write_bindings_to_disk` to separate the generation and writing phases.

Rather than holding the entire output in-memory, then giving it to a function that may decide to write it to disk anyway, I think a more appropriate model would be to use some callback function/function object that is called with each output file's name and content at the point of generation. This would allow us to write them to disk or serialise them into an HTTP response, or whatever else we might decide to do (e.g. pipe through a source code formatter, ...)

I'll create some tickets in the ffig repo to capture this work.

ajbennieston commented:
Ah, having finished reading the code in this PR, I see you're re-implementing the `run` method from `FFIG.py` to avoid the `write_bindings_to_disk` part. Fair enough; in that case it would simply be a matter of a second (or more general) `build_model_from_source`.

That said, I think there's value in the more general approach, especially as it would involve less replicated code in this project. We should be providing a fairly stable external interface that can be used for projects like ffig-explorer, rather than those projects having to hook into the deep internals. In other words, ffig-explorer shouldn't have to call `ffig.generators.render_api_and_obj_classes` directly!

ajbennieston commented:
I would be interested in @jbcoe's thoughts on this...

petr-tik commented:
Awesome - thanks for your ideas. I originally opened the PR to start the conversation on what kind of interface ffig needs to expose and ffig-rest-api should call. I am aware it's hacky and duplicates too much from ffig internals, which makes it prone to future changes. 

What if we add a method to ffig that returns a dictionary of filename: longstring_of_generated_source_file? 

The rest-api will call that method and a front end can serve them as necessary (multiple files in multiple panes).

ajbennieston commented:
That's certainly the easiest approach; but I think there's scope here to make a more general interface by providing a function object that handles `write(filename, generated_src)`. For the main FFIG use cases, that would just write to disk. For the REST API, it would add to a dict, and for other use cases it might do other things, like pipe through `clang-format` or `autopep8`. It would be great if these were chainable, so we could do something like

```python
output_pipeline = [SourceFormatter('py3'), FileWriter()]
generate_bindings(model, template, output_pipeline)
```

petr-tik commented:
Thanks for your help - adding a general FFIG->format->output interface would be useful for ffig-explorer and a CLI

ajbennieston commented:
https://github.com/FFIG/ffig/pull/398 adds a minimal feature to build a `cppmodel` from a string, which you could use immediately in the approach this PR already uses.

The work to properly engineer the FFIG pipeline will take a little longer, so this should help unblock you in the meantime.

petr-tik commented:
great - thanks for moving so quickly. 

Tomorrow is a quieter day for me and I want to submit something minimally functional over the weekend

ajbennieston commented:
Awesome!

petr-tik committed:
Tidy-up according to PR code review

petr-tik committed:
Use the new FFIG build_from_source way of passing unsaved_files

Add bindings to the request form

Turn response into json and print the result

jbcoe commented:
Apologies for delay in responding. I'd go for a 'spike and stabilise' approach where we get something (just) working end-to-end in place. We can come back and engineer it once it works.

jbcoe reviewed (commented):
Do we have any mechanism for (automated) testing this code?

petr-tik commented:
No testing yet, have an item though: #7

petr-tik commented:
The server should work as of last commits. 

I spent yesterday researching web-embeddable text editors, found [ace editor](https://ace.c9.io/) that supports syntax highlighting and is easy enough to include in js. 

Should we add tests to this, merge it and I can continue work on the editor further upstream?

jbcoe commented:
Given that we have nothing in place, as much as I'd like to see tests first I'm happy to approve this so we can iterate upon something concrete.

petr-tik committed:
Added the expected binding, which ffig is supposed to generate

Test the API with a known input-binding pair from main ffig repo

Inconsistency between binutil diff and difflib in Python

```
$ python3 requests_at_explorer.py > exp_bind
$ diff -s exp_bind expected_binding.py
Files exp_bind and expected_binding.py are identical
```

petr-tik committed:
Assert at the bottom of requests_at_explorer now passes

Using difflib.Differ to compare 2 lists of strings

petr-tik commented:
Please review and let me know. This is very much a spike, but we do have an assert at the bottom of requests_at_explorer.py, which checks the diff between expected and API-produced-binding

jbcoe commented:
This is (by necessity) a large change. I’ll take a look on the train and make sure I know what’s going on.

jbcoe commented:
As it stands I can't get this to work (I suspect it's close though).

I run:

```
docker build -t ffig/web-base:latest .
docker run -p 5000:5000 ffig/web-base:latest
```

And get a message from flask which suggests I'm on the right track:

```
WARNING:werkzeug: * Debugger is active!
```

I then run

```
python3 requests_at_explorer.py
```

Which gives no output.

Being a suspicious sort, I added an assertion to check the return code of the REST request. It's not OK so the response content is never checked. 

I suspect all we need is slightly improved instructions and an assertion. if we can get this to pass and fail as expected I'm happy to approve.

jbcoe commented on requests_at_explorer.py:
```diff
@@ -1,31 +1,41 @@
 r = requests.post(
     "http://127.0.0.1:5000/api/gen_bindings_from_tu", data=payload)
-print(r.text)
+if r.status_code == requests.codes.ok:
```
I think we need to check that the status code was ok, if not then something's up and we should not fail silently.

jbcoe commented:
It turns out that reviewing server code on the train (where the web connection is not so hot) does not work so well.

I've added #18 which makes some very minor changes and gets things working nicely enough.

jbcoe committed:
Minor tweaks to get things running (#18)

jbcoe commented:
@petr-tik I'll let you tidy up the commit message and squash-merge.

Maybe see some of @ajbennieston 's commit messages to ffig for some examples.

petr-tik commented:
Will do - thanks
